# 3. Importing Liabraries

In [9]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
path = r'C:\Users\swapn\Instacart Basket Analysis 06_24'

In [7]:
df = pd.read_csv( r'C:\Users\swapn\Instacart Basket Analysis 06_24\Data\Original Data\customers.csv', index_col = False)

In [11]:
df.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

# 4. Wrangle the data , rename columns  and drop columns  add anything to your analysis"
   

In [13]:
df.rename(columns={ 'First Name':'first_name','Surnam':'surname','STATE':'State','n_dependants':'dependants',
              'fam_status':'family_status'}, inplace=True)

In [15]:
df.columns

Index(['user_id', 'first_name', 'surname', 'Gender', 'State', 'Age',
       'date_joined', 'dependants', 'family_status', 'income'],
      dtype='object')

In [19]:
df.head()

user_id first_name   surname  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants family_status  income  
0           3       married  165665  
1           0        single   59285  
2           2       married   99568  
3           0        single   42049  
4           1       married   40374

# 5.Complete the fundamental data quality and consistency check

In [22]:
duplicated = df[df.duplicated()]


In [24]:
duplicated

Empty DataFrame
Columns: [user_id, first_name, surname, Gender, State, Age, date_joined, dependants, family_status, income]
Index: []

# No duplicates are present in customer data 

In [28]:
df.shape

(206209, 10)

In [26]:
# Missing
df.isnull().sum()

user_id              0
first_name       11259
surname              0
Gender               0
State                0
Age                  0
date_joined          0
dependants           0
family_status        0
income               0
dtype: int64

In [36]:
df[ df['first_name'].isnull()]

user_id first_name   surname  Gender         State  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        dependants family_status  income  
53               2       married   41709  
73               0        single   82518  
82               3       married  117099  
99               1       married  155673  
105              3       married  151819  
...            ...           ...     ...  
206038           3       married   87783  
206044           2       married  108488  
206090           1       married   45275  
206162           0        single   56325  
206171           2       married   59222  

[11259 rows x 10 columns]

###### There are 11259 missing values ,located in the 'first name' column, they will be kept as they are.

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        206209 non-null  int64 
 1   first_name     194950 non-null  object
 2   surname        206209 non-null  object
 3   Gender         206209 non-null  object
 4   State          206209 non-null  object
 5   Age            206209 non-null  int64 
 6   date_joined    206209 non-null  object
 7   dependants     206209 non-null  int64 
 8   family_status  206209 non-null  object
 9   income         206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [40]:
df['date_joined']=pd.to_datetime(df['date_joined'])


In [42]:
df.dtypes

user_id                   int64
first_name               object
surname                  object
Gender                   object
State                    object
Age                       int64
date_joined      datetime64[ns]
dependants                int64
family_status            object
income                    int64
dtype: object

## 6.& 7. Combine your customer data with the rest of your prepared Instacart data

In [58]:
df_large = pd.read_pickle(r'C:\Users\swapn\Instacart Basket Analysis 06_24\Data\Prepared Data\ords_prods_merged_grouped.pkl')

In [70]:
df_large_2 = pd.read_pickle(r'C:\Users\swapn\Instacart Basket Analysis 06_24\Data\Prepared Data\ords_prods_merged_derived.pkl')

In [62]:
    df_large.columns


Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'add_to_cart_order', 'reordered', 'user_id', 'eval_set',
       'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order', 'max_order', 'loyalty_flag', 'average_spend',
       'spender_flag', 'customer_frequency', 'frequency_flag'],
      dtype='object')

In [72]:
df_large_2.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'add_to_cart_order', 'reordered', 'user_id', 'eval_set',
       'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order', 'busiest_day', 'busiest_days',
       'busiest_hours'],
      dtype='object')

In [64]:
df_merged = df_large.merge(df,on='user_id',indicator=True)

In [66]:
df_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  add_to_cart_order  reordered  user_id eval_set  ...  first_name  \
0      1107                  7          0    38259    prior  ...         NaN   
1      5319                  3          1   196224    prior  ...       Roger   
2      7540                  4          1   138499    prior  ...      Andrea   
3      9228                  2          0    79603    prior  ...    Nicholas   
4      9273                 30          0    50005    prior  ...        Carl   

     surname  Gender                 State  Age date_joined  dependants  \
0       Bean  Female  District of Columbia   68  2019-06-30           2   
1      Casey    Male              New York   48  2018-04-29           2   
2    Summers  Female         Massachusetts   31  2017-03-11           2   
3      Logan    Male             Tennessee   65  2019-10-27           1   
4  Mcfarland    Male            New Jersey   75  2019-04-09           3   

  family_status  income _merge  
0       married  142744   both  
1       married   43906   both  
2       married   78505   both  
3       married   51010   both  
4       married  115468   both  

[5 rows x 30 columns]

## 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task


In [84]:
df_merged.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_combined_2.pkl'))
